src/include/migraphx/op/rand_uniform.hpp

```diff
@@ -0,0 +1,117 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/**
+ * Random Uniform distribution operator.  Given a shape, populate it with random
+ * values.  Calls to rand_uniform using the same randomization seed will
+ * always generate the same pseudo-random sequence.  Seed can
+ * be given as a runtime argument containing a single value, or a compile-time
+ * attribute.
+ *
+ *      Inputs:   (1) randomization seed (uint32)
+ *                (2) the shape of the set to be populated.
+ *
+ *
+ *      Attributes:  none
+ *
+ *      Output:   Same shape.
+ *
+ */
+#ifndef MIGRAPHX_GUARD_OPERATORS_RAND_UNIFORM_HPP
+#define MIGRAPHX_GUARD_OPERATORS_RAND_UNIFORM_HPP
+
+#include <migraphx/check_shapes.hpp>
+#include <migraphx/argument.hpp>
+#include <migraphx/par_for.hpp>
+#include <migraphx/reflect.hpp>
+#include <random>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace op {
+
+struct rand_uniform
+{
+    // The rand_uniform operation does not contain a random number generator seed
+    // as a member, and expects it to be passed as a runtime input.
+
+    // todo:  not currently settable
+    float range_min = 0.0f;
+    float range_max = 1.0f;
+
+    // todo:  integer data type(s) not yet supported
+    shape::type_t dtype = shape::type_t::float_type;
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.dtype, "dtype"));
+    }
+
+    /**
+     *   Input 1:  seed
+     *   Input 2:  output shape
+     */
+    std::string name() const { return "rand_uniform"; }
+    shape compute_shape(std::vector<shape> inputs) const
+    {
+        check_shapes{inputs, *this, true}.has(2);
+
+        if(inputs.front().type() != shape::type_t::uint32_type)
+            MIGRAPHX_THROW("RAND_UNIFORM:  Input 2 (seed) must have type unsigned int");
+        auto s = inputs.at(1);
+        if(s.dynamic())
+        {
+            return s.with_type(dtype);
+        }
+        else
+        {
+            return s.with_lens(s.lens()).with_type(dtype);
+        }
+    }
+
+    argument compute(const shape& output, std::vector<argument> args) const
+    {
+        // Output goes into the passed buffer, not the shape output
+        (void)output;
+        argument result{args[1].get_shape()};
+
+        uint32_t local_seed = args[0].at<uint32_t>(0);
+
+        std::mt19937 gen(local_seed);
+        std::uniform_real_distribution<> dis(range_min, range_max);
```

Comment:
You need to use `uniform_int_distribution` for when its an integer. Ideally this should probably be moved into the visitor:

```cpp
        result.visit([&](auto output) {
            using type = typename decltype(output)::type;
            if constexpr(std::is_integral<type>{})
            {
                std::uniform_real_distribution<type> dis;
                std::generate(output.begin(), output.end(), dis);
            }
            else
            {
                std::uniform_int_distribution<type> dis;
                std::generate(output.begin(), output.end(), dis);
            }
        });
```

Reply:
This PR has been full of comments for some time stating that we need to decide whether this operation needs to support integer data types at this time.  I take it from this comment that you think it should.  This will require me to add range arguments such as min and max, since discrete distributions don't exist without a range and the default range of (0, 1) can't be used with integers.  I'll also have to add tests for the new arguments.  I don't know if we have a use case at hand that requires the availability of discrete random distributions.  Work on the Multinomial op and dynamic batch input testing will be blocked for an additional day or two while I carry out this work.

Reply:
> This PR has been full of comments for some time stating that we need to decide whether this operation needs to support integer data types at this time. I take it from this comment that you think it should.

The code you wrote runs on integers except it will compute incorrect results. Its simple enough to switch to `uniform_int_distribution` for integers. Plus, its good for reference operators to have coverage for all input types. 

> This will require me to add range arguments such as min and max, since discrete distributions don't exist without a range and the default range of (0, 1) can't be used with integers.

I agree, but I dont think this needs to be implemented in this PR. You can just use the default constructor for the distribution classes which will pick (0, 1) for floats and (0, max) for integers. 

> I don't know if we have a use case at hand that requires the availability of discrete random distributions.

If the model is quantized it could very likely use integral distributions. 

> Work on the Multinomial op and dynamic batch input testing will be blocked for an additional day or two while I carry out this work.

Sure, we still need to merge in Charlie's allocation PR anyways, and there are other comments in the PR that needs to be addressed.

Reply:
That line with decltype doesn't compile.  Is there another way to get the data type of a tensor_view?  The following doesn't give a compilation error but gives a wrong result:


```
             using type =  decltype((output[0]));

            if constexpr(std::is_integral<type>{})
...
```

The error message in this case indicates the decltype is a signed char &